react/getting_started_with_react/rendering_techniques.md

```diff
@@ -0,0 +1,288 @@
+
+We are not limited to rendering an element or nothing with `null` we can also decide whether to render an element or another element.
+
+#### Other Ways to Render Conditionally
+
+We can also use `if`, `if/else`, and `switch` to conditionally render something.
+
+This time we will remove all the animals from the list, and we will also have two conditions:
+
+1.  Check if the `animals` property is provided
+2.  Check if the `animals` length is greater than 0
+
+We will frequently be dealing with lists in the future, and we also need to consider what to render if the list is empty or does not exist at all. You certainly would not want to see a blank page, would you? Let us try to implement that:
+
+~~~javascript
+function List(props) {
+  if (!props.animals) {
+    return <div>Loading...</div>;
+  }
+
+  if (props.animals.length === 0) {
+    return <div>There are no animals in the list!</div>;
+  }
+
+  return (
+    <ul>
+      {props.animals.map((animal) => {
+        return <li key={animal}>{animal}</li>;
+      })}
+    </ul>
+  );
+}
+
+function App() {
+  const animals = [];
+
+  return (
+    <div>
+      <h1>Animals: </h1>
+      <List animals={animals} />
+    </div>
+  );
+}
+~~~
+
+In our `<List />` component, we have two `if` statements acting as a guard that immediately returns an element based on the condition.
+
+One is to check if the property `animals` exists, and the other is to check if the length of the list is greater than 0. In this case, our list is empty, so the second if statement executes, it will immediately return the `<div>` element that contains the text "There are no animals in the list"
+
+If we remove the `animals` property:
+
+~~~javascript
+function App() {
+  const animals = [];
+
+  return (
+    <div>
+      <h1>Animals: </h1>
+      <List />
+    </div>
+  );
+}
+~~~
+
+The first `if` statement will now execute and return a `<div>` with the text "Loading..." This is often the case when you are fetching from an API, since it might take some time to actually retrieve the data, it is good practice to show an indicator for that.
+
+If none of those checks passed, then we have the data we need to render the list successfully. Try it out by adding items to the `animals` list and adding the property back.
+
+You can, Of course, also accomplish this with the ternary and && operators.
+
+~~~javascript
+function List(props) {
+  return (
+    <ul>
+      {props.animals.map((animal) => {
+        return <li key={animal}>{animal}</li>;
+      })}
+    </ul>
+  );
+}
+
+function App() {
+  const animals = [];
+
+  return (
+    <div>
+      <h1>Animals: </h1>
+      {animals.length > 0 ? (
+        <List animals={animals} />
+      ) : (
+        <div>There are no animals in the list!</div>
+      )}
+    </div>
```

Comment:
I was thinking it makes sense to perform the checks inside the `List` component, because we can check for the existence of a property of an object. However, if we, for example, have the checks inside the `App` component, and then remove the `animals` variable, it will throw an error. Since I was thinking of `App` as the parent component that has the data, and somehow slightly introduce passing down of props.

And yes, I believe we can make the `ternary` and `&&` examples to be consistent with the previous `if` example. But there is one change for the `&&` example to work in the `List` component, we will be needing to change the enclosing element to a React fragment or just a `div`, I believe it will be introduced in the JSX lesson(?)
```js
function List(props) {
  return (
    <>
      {!props.animals && <div>...</div>}
      {props.animals && props.animals.length > 0 && <ul>...</ul>}
      {props.animals && props.animals.length === 0 && <div>...</div>}
    </>
  )
}
```
It's quite intimidating to look at, but would love to know more of your thoughts about this. Thank you.

Reply:
I agree with what you have here, and React fragments are fine.